mimic std::optional for flagged Waypoint fields.

tsteven4 opened the pull request:
This implementation preserves the use of wpt_flags, which saves considerable space for every Waypoint instance compared to std::optional.

The Waypoint fields that use wpt_flags are made private to force usage of the member functions that access them.  This should guarantee that the wpt_flags are kept in sync with the data fields, and unset data fields are not used.

tsteven4 committed:
mimic std::optional for flagged Waypoint fields.

robertlipe commented on defs.h:
```diff
@@ -405,27 +406,27 @@ class Waypoint
-  float speed;   	/* Optional: meters per second. */
   fix_type fix;	/* Optional: 3d, 2d, etc. */
   int  sat;	/* Optional: number of sats used for fix */
 
```
Can't most of these (dop, cadence,  ⁠ಠpower, heart rate, etc).use the same scheme?

tsteven4 replied:
we aren't consistent.  some things assume zero indicates invalid.  I think heart rate, cadence and power all use this convention.  It looks like the *dops used >0 or ==0.  we could convert them to use the new macros for consistency, it shouldn't cost too much as the macros are all inline (as they are defined in class).  perhaps this is best left to another PR, I didn't even review the changes in this one myself yet, I was rushing to get a draft out you could look at.

tsteven4 committed:
manual adjustments to waypt access usage.

tsteven4 commented on garmin_txt.cc:
```diff
@@ -711,15 +711,15 @@ track_disp_wpt_cb(const Waypoint* wpt)
+    float temp = (wpt->temperature_value_or(-999));
     if (temp != -999) {
       print_temperature(temp);
     }
```
shouldn't the printing of temperature, and it's tab, be outside the if statement?

tsteven4 committed:
split optional Waypoint flags, tune packing.

tsteven4 committed:
bring op_flags into Waypoint.

tsteven4 committed:
add comments about future Waypoint optional fields.

tsteven4 committed:
sort Waypoint class almost normally.

private members refered to with decltype need to be defined
before they are referred to.

tsteven4 committed:
remove extra parens around *has_value().

tsteven4 committed:
remove extra parens around *value_or().

tsteven4 committed:
remove more extra parens.
